python/grass/jupyter/timeseries.py

```diff
@@ -0,0 +1,274 @@
+# MODULE:    grass.jupyter.timeseries
+#
+# AUTHOR(S): Caitlin Haedrich <caitlin DOT haedrich AT gmail>
+#
+# PURPOSE:   This module contains functions for visualizing raster and vector
+#            space-time datasets in Jupyter Notebooks
+#
+# COPYRIGHT: (C) 2021 Caitlin Haedrich, and by the GRASS Development Team
+#
+#           This program is free software under the GNU General Public
+#           License (>=v2). Read the file COPYING that comes with GRASS
+#           for details.
+"""Create and display visualizations for space-time datasets."""
+
+import tempfile
+import os
+import weakref
+import grass.script as gs
+from .display import GrassRenderer
+
+
+def collect_lyr_dates(timeseries, etype):
+    """Create lists of layer names and start_times for a
+    space-time raster or vector dataset.
+
+    For datasets with variable timesteps, makes step regular with
+    "gran" method for t.rast.list or t.vect.list then fills in
+    missing layers with previous timestep layer. If first time step
+    is missing, uses the first available layer.
+
+    :param str timeseries: name of space-time dataset
+    :param str etype: element type, "stvds" or "strds"
+    """
+    if etype == "strds":
+        rows = gs.read_command(
+            "t.rast.list", method="gran", input=timeseries
+        ).splitlines()
+    elif etype == "stvds":
+        rows = gs.read_command(
+            "t.vect.list", method="gran", input=timeseries
+        ).splitlines()
+    else:
+        raise NameError(
+            _("Dataset {} must be element type 'strds' or 'stvds'").format(timeseries)
+        )
+
+    # Parse string
+    new_rows = [row.split("|") for row in rows]
+    new_array = [list(row) for row in zip(*new_rows)]
+
+    # Collect layer name and start time
+    for column in new_array:
+        if column[0] == "name":
+            names = column[1:]
+        if column[0] == "start_time":
+            dates = column[1:]
+
+    # For variable timestep datasets, fill in None values with
+    # previous time step value. If first time step is missing data,
+    # use the next non-None value
+    for i, name in enumerate(names):
+        if name == "None":
+            if i > 0:
+                names[i] = names[i - 1]
```

Comment:
I think the filling should be optional. In your example when you unregister one map, it feels like the user should know the gap is there, the behavior seems misleading. But I would be interested in other opinions about that.

Reply:
I wanted to say this too, but was unsure of what happens in g.gui.animate for example (as to compare), when a date is missing in a time series with regular intervals. If I understand the current behavior in the notebook correctly, filling a missing date with the map from the previous date would be wrong IMHO... would it be possible to leave it blank? Maybe not nice but it would be correct as there's no map for that date, otherwise we would be making assumptions about the data (indeed, another possibility would be to interpolate, but again we would be making assumptions). I darkly remember to have created vector data animations where there were missing vector maps for certain dates and the user just see the time passing (in the time stamps) but no vector map was displayed until there were dates with data again.

Reply:
I also wasn't sure that this behavior was ideal but wasn't sure of better alternatives. I think having it be an option makes a lot of sense, maybe something like:

`gj.TimeSeries( ... , gapfill=True,  method="previous" or "next")`. If `gapfill=False`, it returns a blank image?

Interpolation seems complex enough that it should be left to the user to interpolate the gaps and pass the filled dataset.  I do think it makes sense to offer the option to "fill gap with previous or next map"  since otherwise, some sparse and irregularly-spaced datasets will create visuals that are mostly blank (i.e. see [Nags Head example](http://ncsu-geoforall-lab.github.io/grass-temporal-workshop/))

Reply:
If gapfill and method are needed, I would collapse that into `gap_fill={method}` with method being `"previous"`, `"next"`, `None`.